### DISCUSSION TITLE
Silent crash in react-native when used with react-native-gesture-handler

### DISCUSSION THREAD

**Discussion by siarhei-klimuts** [pmndrs/zustand#2603] [Bug report] [answered]
I have a code like this:
```
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import { create } from "zustand";

const useGameState = create((set) => ({
    move: 0,
    delta: 0,
    pos: new Vector3(0, 0, 0),
    setMove: (move) => set(() => ({ move })),
    setDelta: (delta) => set(() => ({ delta })),
    setPos: (pos) => set(() => ({ pos })),
  }));
  
function Component() {
    const setMove = useGameState(({ setMove }) => setMove);

    const pan = Gesture.Pan()
    .minDistance(0)
    .onStart(() => {
      try {
        setMove(1);
      } catch (e) {
        console.log("!", JSON.stringify(e));
      }
      }
    })
    .onUpdate(() => {
        ...some code
    })
    .onEnd(() => {
        ...some code
    });

    return (
        <GestureDetector gesture={pan}>
            ...some views
            ...some modals
            ...some r3f
        </GestureDetector>
    )
}
```

Calling `setMove` in `Gesture.Pan().onStart` function causes my app to crash in android expo go app (dev mode).
Tried to set state with `useGameState.setState({ move: 1 })` - the same result. In other code `setDelta` works fine.
Expo web build (dev) works great, no issue there.

The error object logged to console is `{}`

my deps:
"dependencies": {
    "@expo-google-fonts/roboto": "^0.2.3",
    "@expo/metro-runtime": "~3.2.1",
    "@react-native-async-storage/async-storage": "1.23.1",
    "@react-spring/three": "^9.7.3",
    "@react-three/drei": "^9.106.0",
    "@react-three/fiber": "^8.16.6",
    "date-fns": "^3.6.0",
    "expo": "~51.0.9",
    "expo-asset": "~10.0.6",
    "expo-av": "~14.0.5",
    "expo-constants": "~16.0.2",
    "expo-gl": "~14.0.2",
    "expo-linking": "~6.3.1",
    "expo-router": "~3.5.15",
    "expo-status-bar": "~1.12.1",
    "immer": "^10.1.1",
    "lodash-es": "^4.17.21",
    "nativewind": "4.0.36",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-native": "0.74.1",
    "react-native-game-engine": "^1.2.0",
    "react-native-gesture-handler": "~2.16.1",
    "react-native-heroicons": "^4.0.0",
    "react-native-reanimated": "^3.10.0",
    "react-native-safe-area-context": "4.10.1",
    "react-native-screens": "3.31.1",
    "react-native-web": "~0.19.10",
    "tailwind-merge": "^2.3.0",
    "tailwind-variants": "^0.2.1",
    "three": "^0.164.1",
    "zustand": "^4.5.2"
  },

Can't provide more information as I just installed zustand and it looks like it does not want to be used by me :)
Happy fixing!

**Comment by dbritto-dev** [comment]
@siarhei-klimuts replace this `const setMove = useGameState(({ setMove }) => setMove);` with `const setMove = useGameState(state => state.setMove);`

**Reply by siarhei-klimuts** [reply]
> @siarhei-klimuts replace this `const setMove = useGameState(({ setMove }) => setMove);` with `const setMove = useGameState(state => state.setMove);`

No difference

**Reply by dbritto-dev** [reply, marked as answer]
@siarhei-klimuts as far I see should something like this:

```ts
const pan = Gesture.Pan()
    .minDistance(1)
    .onStart(() => {
      setMove(() => 1)
    })
    .onUpdate((event) => {
      setMove(current => current + 1)
    })
    .runOnJS(true);
```

But your implementation looks different

**Reply by dbritto-dev** [reply]
Ref:

```ts
const pan = Gesture.Pan()
    .minDistance(1)
    .onStart(() => {
      prevTranslationX.value = translationX.value;
      prevTranslationY.value = translationY.value;
    })
    .onUpdate((event) => {
      const maxTranslateX = width / 2 - 50;
      const maxTranslateY = height / 2 - 50;

      translationX.value = clamp(
        prevTranslationX.value + event.translationX,
        -maxTranslateX,
        maxTranslateX
      );
      translationY.value = clamp(
        prevTranslationY.value + event.translationY,
        -maxTranslateY,
        maxTranslateY
      );
    })
    .runOnJS(true);
```

**Reply by siarhei-klimuts** [reply]
> ```ts
> .runOnJS(true);
> ```

Thank you! Adding `.runOnJS(true)` fixed the issue.

So the only thing left to notice is `Gesture.Pan()` worked fine without `.runOnJS(true)` until I tried to call zustand code with it.

[Here](https://docs.swmansion.com/react-native-gesture-handler/docs/gestures/pan-gesture#runonjsvalue-boolean) is the documentation for this function.

**Reply by dbritto-dev** [reply]
@siarhei-klimuts I believe that you need use runOnJs since those calls should run on js thread not ui thread
